internal/queue/consumer.go

```diff
@@ -0,0 +1,285 @@
+/*
+ * Copyright 2025 Canonical Ltd.
+ * See LICENSE file for licensing details.
+ *
+ * Package queue provides AMQP consumer implementations for job events.
+ *
+ */
+
+package queue
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"log/slog"
+	"strconv"
+	"sync"
+	"time"
+
+	"github.com/canonical/github-runner-operators/internal/database"
+	amqp "github.com/rabbitmq/amqp091-go"
+)
+
+const platform = "github"
+
+// JobDatabase is a small interface that matches the relevant method on internal/database.Database.
+type JobDatabase interface {
+	AddJob(ctx context.Context, job *database.Job) error
+	UpdateJobStarted(ctx context.Context, platform, id string, startedAt time.Time, raw map[string]any) error
+	UpdateJobCompleted(ctx context.Context, platform, id string, completedAt time.Time, raw map[string]any) error
+}
+
+// AmqpConsumer is an AMQP consumer for workflow job events.
+type AmqpConsumer struct {
+	amqpChannel    amqpChannel
+	amqpConnection amqpConnection
+	connectFunc    func(uri string) (amqpConnection, error)
+	uri            string
+	queueName      string
+	db             JobDatabase
+	mu             sync.Mutex
+	logger         *slog.Logger
+}
+
+// NewAmqpConsumer creates a new AmqpConsumer with the given dependencies.
+func NewAmqpConsumer(uri, queueName string, db JobDatabase, logger *slog.Logger) *AmqpConsumer {
+	if logger == nil {
+		logger = slog.Default()
+	}
+	return &AmqpConsumer{
+		uri:         uri,
+		queueName:   queueName,
+		connectFunc: amqpConnect,
+		db:          db,
+		logger:      logger,
+	}
+}
+
+// Start begins consuming messages from the AMQP queue and processing them.
+func (c *AmqpConsumer) Start(ctx context.Context) error {
+	err := c.resetConnectionOrChannelIfNecessary()
+	if err != nil {
+		return err
+	}
+
+	msgs, err := c.consumeMsgs()
+	if err != nil {
+		return err
+	}
+
+	c.logger.Info("AMQP consumer started", "queue", c.queueName)
+
+	for {
+		select {
+		case <-ctx.Done():
+			return ctx.Err()
+		case msg, ok := <-msgs:
+			if !ok {
+				return fmt.Errorf("messages channel closed")
+			}
+			if err := c.handleMessage(ctx, msg); err != nil {
+				c.logger.Error("failed to handle message", "error", err)
+			}
+		}
+	}
+}
+
+// WorkflowJobWebhook represents the structure of a workflow job webhook message.
+type WorkflowJobWebhook struct {
```

Comment:
use this library for parsing webhook https://pkg.go.dev/github.com/google/go-github/github#ParseWebHook

Reply:
but then how would the switch statement works in this case? from the documentation, it seems to work for Commit/Create/PushEvent, but im not sure if action such as "queued", "in_progress", etc. can be parsed?
```
  switch event := event.(type) {
  case *github.CommitCommentEvent:
      processCommitCommentEvent(event)
  case *github.CreateEvent:
      processCreateEvent(event)
  ...
  }
  ```

Reply:
@florentianayuwono I created a small example of switching over the actions. 

```go
	event, err := github.ParseWebHook(headers["X-GitHub-Event"], payload)
	if err != nil {
		panic(err)
	}
	jobEvent, ok := event.(*github.WorkflowJobEvent)
	if !ok {
		fmt.Println("event is not a workflow_job, discarding")
		// do something
		return
	}
	switch jobEvent.GetAction() {
	case "queued":
		// do something
	case "waiting":
		// do something
	case "in_progress":
		// do something
	case "completed":
		// do something
	}
```